### ISSUE TITLE
Error handling: Don't call exit(n) inside libtesseract

### ISSUE THREAD

**Issue by amitdo** [tesseract-ocr/tesseract#2990]
https://github.com/tesseract-ocr/tesseract/blob/01535706ecd11a495f9f6e65ee0118b3537a115a/src/textord/strokewidth.cpp#L121-L128

https://github.com/tesseract-ocr/tesseract/blob/2a37f5dd6226156bf32e60d1ae0209a383af4843/src/classify/intproto.cpp#L231-L238

https://github.com/tesseract-ocr/tesseract/blob/912c9978dacf9f7a7cc97ba9a906b40e3df8fc09/src/dict/permdawg.cpp#L111-L121

**Comment by stweil** [comment]
Are there examples which trigger those exits? That would help testing fixes this issue.

**Comment by amitdo** [comment]
No, I don't have examples which trigger those exits.

**Comment by jbarlow83** [comment]
@stweil Any call to `exit()` from code intended for use in a library is a bug - it doesn't really matter how to trigger them. For a GUI that binds libtesseract, the application will be nuked from orbit by `exit()` and probably won't be able to get an error message to user.

Ideally Tesseract would just throw an `std::exception` because those exits are really deep in the call stack. Are exceptions still forbidden? The standard library can and does throw exceptions, they're already de facto in use. 

If we can't use exceptions, the thing to do is define a function pointer that defaults to `exit()`...
```c++
void (*tesseract_exit)(int error_code) = exit;
```
...and add an API call that lets the application provide its own exit handler. That way the application can throw an exception or do whatever makes sense in context.

**Comment by amitdo** [comment]
The ban on exception came from Google coding rules. Since Ray leaved this project  years ago, we don't have to be limited by their internal coding standards. 

@stweil, should we use exceptions instead of these exit() calls?

**Comment by jbarlow83** [comment]
C++ exceptions are zero cost when the exception does not occur.

**Comment by bertsky** [comment]
> Calling `exit` or maybe better `abort` in case of an internal software bug is fine for me.

Calling `abort()` **in a library** is itself as much a bug as the original `exit()` formulation. The application will be gone along with the library and can not even do anything about it.

This issue is also much bigger IMO: it's not just these few places, but every `ASSERT_HOST` (any there are plenty) delegates to `abort()` as well. (Or, if compiled with `-g`, even to a segfault event.): https://github.com/tesseract-ocr/tesseract/blob/main/src/ccutil/errcode.cpp#L72-L83

**Comment by stweil** [comment]
I disagree. Assertions are useful to find software bugs. It's perfectly fine that they abort, trigger a debugger breakpoint or a core dump. Like this, chances that they get fixed are much higher, so ideally no assertions will trigger an abort in the future. Currently there are [5 open issues which are related to an unexpected termination](https://github.com/tesseract-ocr/tesseract/issues?q=is%3Aissue%20state%3Aopen%20assert%20label%3A%22unexpected%20termination%22), but it looks like they are based on only two real code issues. Most users are not affected by these issues, so for me they have low priority. And the faulty code which triggers the assertions is buggy and must be fixed, not the assertions that abort the program.

There are of course other places for which any kind of program termination must be considered as a bug. An API call for example should never trigger an assertion. The examples given in the original issue report here are such examples. But even in these (rare) cases calling exit can be accepted as an intermediate solution as long as there is no better fix which handles the runtime errors in a better way. Sometimes fixes cost a lot of resources. And this project is currently driven by volunteers which are free to decide where they spend their personal resources.

**Comment by bertsky** [comment]
@stweil you missed my point entirely. I am (of course!) **not** advocating against assertions. The question is what they do / how they are implemented.

The OP is talking about Tesseract as a library. That's why I was picking up the issue. For me it's obviously the Python side of that C library I am interested in. That's our context, ok? 

So: Code called by any part of the Tesseract API simply should **never** `abort` or `exit` (or worse: segfault), because that is itself a severe bug of that library, in addition to the assertion failure that triggered it (however mild or severe it was).

It is totally irrelevant how many **currently known** open issues are causing such aborts. My case does not currently seem to be tracked by an issue (but it also shows up as a failed pageres iterator assert, as some of these also showed). 

The point was: because `abort` (or `exit` or deliberate segfault) in a library makes error handling impossible for applications, it must not be employed, ever. 

(For example, in ocrd_tesserocr, error handling simply cannot happen on any level. So the processor or server will be gone without logging or leaving a well-defined state of the workspace.)

C++ exceptions have been suggested already. I am not qualified enough to assess if that's itself problematic, or whether they may be other options.

**Comment by stweil** [comment]
I don't think that I missed your point, but still disagree. The only reasonable error handling for buggy code is debugging. If ocrd_tesserocr or any other component of OCR-D terminates unexpectedly, this is a severe bug which should be easy to fix as long as it can be reproduced. Simply enable core dumps. Then nothing will be gone without extensive logging!

**Comment by MehmedGIT** [comment]
> C++ exceptions are zero cost when the exception does not occur.

Here is also an [article](https://mmomtchev.medium.com/the-true-cost-of-c-exceptions-7be7614b5d84) with some test data to support that statement. From the article:

> If you feel tempted to disable exceptions when building your C++ project, think again. Take the time to actually test it. Check if actually improves anything, because it definitely breaks things. If it is a standalone application, then this will probably affect only yourself — when you start debugging it. But if you are building a target that people will be linking with, I very strongly recommend you to avoid doing this. Otherwise, sooner or later, you will have an user that will spend countless hours debugging a very bizarre problem."

So, please reconsider the decision not to have exceptions. Even if there may be a small performance trade-off.

> @bertsky
> The point was: because `abort` (or `exit` or deliberate segfault) in a library makes error handling impossible for applications, it must not be employed, ever.
> 
> (For example, in ocrd_tesserocr, error handling simply cannot happen on any level. So the processor or server will be gone without logging or leaving a well-defined state of the workspace.)

Completely agree with this point.
